### PR TITLE
Rule 3.3

### LINKED ISSUE #56: [Rule 3.3] Braces
Rule 3.3 Braces must be used in conditional statements and loop blocks

  In `if`, `else`, `when`, `for`, `do`, and `while` statements, even if the program body is empty or contains only one statement, braces should be used.
  
**Exception:** *Only* The only exception is ternary operator in Kotlin (it is a single line `if () <> else <>` ) 
When the entire expression can be 

Bad example:

```kotlin
val value = if (string.isEmpty())  // WRONG!
                0
            else
                1
```

Valid example: 

```kotlin
val value = if (string.isEmpty()) 0 else 1  // Okay
```

```kotlin
if (condition) {
    println("test")
} else {
    println(0)
}

### PR DESCRIPTION
**Pull request by petertrr**
This closes #56 

### What's done:
* Moved some rules to packages
* Implemented checkers for rule 3.3
* Added tests
* Updated available-rules.md

### TODO:
* [x] Fixers and tests for them
* [x] Update rules-config.json

### Fixmes:
* Is there any example for `else` with empty body?

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            }
+        }
+
+        if (hasElseBranch && elseNode?.elementType != IF && elseNode?.elementType != BLOCK) {
+            NO_BRACES_IN_CONDITIONALS_AND_LOOPS.warnAndFix(configRules, emitWarn, isFixMode, "ELSE",
+                    (elseNode?.treeParent?.prevSibling { it.elementType == ELSE_KEYWORD } ?: node).startOffset) {
+                if (elseNode != null) {
+                    ifPsi.`else`!!.replaceWithBlock(indent)
+                } else {
+                    // fixme: in which case else can have empty body?
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:95]: simply someone added empty branch? 
for `when` we even have such rule in code style (for when we should always have else branch)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    private fun checkWhenBranches(node: ASTNode) {
+        (node.psi as KtWhenExpression).entries.asSequence()
+                .filter { it.expression != null && it.expression!!.node.elementType == BLOCK }
+                .map { it.expression as KtBlockExpression }
+                .filter { it.statements.size == 1 }
+                .forEach {
+                    NO_BRACES_IN_CONDITIONALS_AND_LOOPS.warnAndFix(configRules, emitWarn, isFixMode, "WHEN", it.node.startOffset) {
+                        it.astReplace(it.firstStatement!!.node.psi)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:135]: don't get it, where do you add braces?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    private fun isSingleLineIfElse(node: ASTNode, elseNode: ASTNode?): Boolean {
+        val hasSingleElse = elseNode != null && elseNode.elementType != IF
+        return node.treeParent.elementType != ELSE && hasSingleElse && node.text.lines().size == 1
+    }
+
+    private fun checkLoop(node: ASTNode) {
+        val loopBody = (node.psi as KtLoopExpression).body
+        val loopBodyNode = loopBody?.node
+        if (loopBodyNode == null || loopBodyNode.elementType != BLOCK) {
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:110]: loopBodyNode  what is the case when it is null?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt (hunk 1 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:95]: Yes, but I couldn't come up with example which can be compiled with empty `else` branch for `if` statement. If you do something like 

```
if (condition)
    foo()
else
```

compiler will complain that expression is expected.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt (hunk 3 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:110]: do-while with empty body (again, couldn't find correct examples for other loops with empty body)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt (hunk 2 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:135]: I don't, for `when` statements we decided to only remove braces for single-line branches, didn't we?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt (hunk 3 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:110]: do-while with empty body (again, couldn't find correct examples for other loops with empty body)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt (hunk 2 above)
Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/BracesInConditionalsAndLoopsRule.kt:135]: ah, yeah, really, forgot it

**Pull request merged**